### PR TITLE
Allow for more than 32 tracks per media type

### PR DESCRIPTION
**Pull request by JoelLinn**
The track selection is handled by a bit mask. 
This mask was 32 bit wide so the maximum number of streamable tracks was limited to 32.
If the video contained more tracks, the logic in [mp4_parser.c:2867](https://github.com/JoelLinn/nginx-vod-module/blob/ce273ec54d12bd2141297370dd95b9e6387dcfcc/vod/mp4/mp4_parser.c#L2867) would wrap and try to stream more than one track at once for indexes greater than 31.

This patch uses an array of `uint64_t` as bitmask.
The maximum number of tracks is selected at compile time. It can be increased by adding `-DNGX_VOD_MAX_TRACK_COUNT=256` to the cc configuration options and defaults to 64. (This mechanism is used by other nginx modules as well, for example lua.)

A number of macros and inline functions was added to the header which already defined the existing bitset macros. They are needed to simplify operations like `(mask1 & mask2) != 0`. With `-O2` or greater they are optimized and the mentioned operation compiles to only 3 x86 instructions (for up to 256 tracks) if vector extensions are enabled (`-mavx2`).

I had to add a typedef for the track mask array in order to avoid array type decay when using multi dimensional arrays as function parameters (like `size_t tracks_mask[MEDIA_TYPE_COUNT][MASK_ARRAY_SIZE]`).

With the python script `generate_many_tracks.py` it's possible to create a mp4 with many language tracks (≈600), which I successfully tested with `NGX_VOD_MAX_TRACK_COUNT=640` and the [dash reference client](http://reference.dashif.org/dash.js/nightly/samples/dash-if-reference-player/index.html) (vlc seems to have a bug with such a high number of tracks). Not that we need that amount of tracks but the test should prove that this PR fixes the underlying issue for good.

Old text:
> This patch widens the masks to 64 bits.
> It also changes `MAX_TRACK_COUNT` to `64` in order to catch unsupported files deterministically.
> I am not 100% sure about changing `MAX_TRACK_COUNT` though, I never looked into mixing - it might be possible that it's more sensible to keep it at `1024` and add another check somewhere else.
> 
> In our use case, we have one source language audio track and two additional tracks for each  language (translation only + translation with quiet source language track). That effectively limits us to only 16 languages.
> 
> **Edit:**
> We might hit the 64 track limit at a later point, but fixing that requires larger changes so this will do for now.
> I would prefer to discuss the needed changes with maintainers before implementation.
>

### CODE REVIEW

**Comment by JoelLinn** [comment]
If we want to support more than 64 tracks, we need a bitset type. There are two approaches:

* A type (plus functions) that exposes bitset functionality on a dynamic number of bits, like https://github.com/goldsborough/bitset but based on the ngx array. Number of tracks is virtually unlimited but branching (loop) is required when working with the bitset (performance impact?)
* Static size bitset type (based on compile time constants). Has no branching (the small constant size for loops get unrolled by the compiler) and compiles to the same assembly as current master if the bitset size is `sizeof(size_t)`.

The dynamic approach is very elegant, especially since there are no templates in C so that the static bitset would probably be a macro mess.

@erankor What are your thoughts on this? Do we need to create some benchmarks?

**Comment by erankor** [comment]
Question is whether we actually need more than 64, it sounds like quite a lot to me... 
If it is needed, I'd go with the static approach. mostly because the dynamic will probably add quite a few error cases to handle. And, there will probably also be complications when working with masks of different sizes. 
I don't think the static approach will be messy, can use the existing vod_is_bit_set/vod_set_bit macros + add some function to test if all bits in a mask are set.

Few small comments regarding the existing PR (most are irrelevant if you implement for >64...) -
1. can replace `(uint64_t)1` with `1ULL`
2. no need for `parse_utils_extract_uint64_token` - the mask is 64, but track indexes can remain 32
3. better add a const for 0xffff..., I should have done it when I wrote it, but with 64 it's more painful :)

**Comment by JoelLinn** [comment]
I agree that most won’t need more than 64 tracks. Static limit it is then👍

**Comment by JoelLinn** [comment]
I revised the existing PR and replaced the track masks with an array (like it has been with the language masks before).
The original comment was updated with the new approach. 

Although it's a lot of changes and additions, the compiler is able to optimise most of it out - especially for a max track count of 64 (which fits in a single register)

**Comment by JoelLinn** [comment]
I rebased and added the bitset test to the new CI pipeline.

**Comment by erankor** [comment]
@JoelLinn, I briefed through the changes, and while the code seems well written, it is more complex that I'd want it to be...
I think we should try to simplify it, I was thinking about 2 things -
1. Assume MAX_TRACK_COUNT is a multiple of 64 - remove all the code that handles residual bits and force it during compilation (#if + #error)
2. Use uint64_t for the mask - I don't really care about 32 bit systems, if that's the reason you used size_t... I want to keep it functioning for the few folks who use them, but if it will run slower there, I'm fine with that. There are a few things that become a bit simpler if you know the size of mask elements.

**Comment by jessp01** [comment]
I agree with @erankor's points. I don't think this code runs on many 32bit archs and I feel these are quite rare in general nowadays.. We shouldn't break functionality of course, but slower exec in favour of cleaner/simpler code flows is a well justified trade off, IMHO.

**Comment by JoelLinn** [comment]
I switched to uint64_t and removed the residual handling. Added an error for `MAX_TRACK_COUNT % 64 != 0`

File: vod/manifest_utils.c [hunk 1]
```diff
@@ -111,23 +134,23 @@ manifest_utils_build_request_params_string_per_sequence_tracks(
...
 		// video tracks
-		if (tracks_mask[MEDIA_TYPE_VIDEO] == 0xffffffff)
+		if (vod_is_all_bits_set(tracks_mask[MEDIA_TYPE_VIDEO], MAX_TRACK_COUNT))
 		{
 			result_size += sizeof("-v0") - 1;
 		}
 		else
 		{
-			result_size += vod_get_number_of_set_bits(tracks_mask[MEDIA_TYPE_VIDEO]) * (sizeof("-v32") - 1);
+			result_size += vod_get_number_of_set_bits_in_mask(tracks_mask[MEDIA_TYPE_VIDEO], MAX_TRACK_COUNT) * (sizeof("-v999") - 1);
```
Comment (erankor) [inline vod/manifest_utils.c:143]: this can be problematic if someone goes crazy and uses MAX_TRACK_COUNT > 1000 + it will allocate unneeded space with the default value of 64. 
I suggest we do something like - 
```
#define MAX_TRACK_INDEX_LEN (sizeof(ngx_value(MAX_TRACK_COUNT)) - 1)
```

File: vod/manifest_utils.c [hunk 2]
```diff
@@ -48,12 +48,34 @@ manifest_utils_write_bitmask(u_char* p, uint32_t bitmask, u_char letter)
 	return p;
 }
 
-static uint32_t*
+static u_char*
+manifest_utils_write_bitmask(u_char* p, uint64_t* bitmask, uint64_t* mask_temp, uint32_t max_bits, u_char letter)
```
Comment (erankor) [inline vod/manifest_utils.c:52]: i think it's worth adding an intermediate function that is specific to track_mask_t, that will not require `mask_temp`

File: vod/media_format.h [hunk 3]
```diff
@@ -14,7 +14,14 @@
 // constants
 #define MAX_CODEC_NAME_SIZE (64)
 #define MAX_FRAME_SIZE (10 * 1024 * 1024)
-#define MAX_TRACK_COUNT (1024)
+#ifdef NGX_VOD_MAX_TRACK_COUNT
+#define MAX_TRACK_COUNT (NGX_VOD_MAX_TRACK_COUNT)
+#else
+#define MAX_TRACK_COUNT (64)
+#endif
+#if MAX_TRACK_COUNT % 64 != 0
```
Comment (erankor) [inline vod/media_format.h:22]: would prefer to put it inside the #ifdef (even though it makes no difference...)

File: vod/media_set_parser.c [hunk 4]
```diff
@@ -432,10 +432,10 @@ media_set_parse_tracks_spec(
 	void* dest)
 {
 	media_filter_parse_context_t* context = ctx;
-	uint32_t* tracks_mask = dest;
+	track_mask_t* tracks_mask = dest;
 	u_char* end_pos = value->v.str.data + value->v.str.len;
 
-	vod_memzero(tracks_mask, sizeof(tracks_mask[0]) * MEDIA_TYPE_COUNT);
+	vod_memzero(tracks_mask, sizeof(track_mask_t) * MEDIA_TYPE_COUNT);
```
Comment (erankor) [inline vod/media_set_parser.c:438]: same as earlier, can revert this change

File: vod/mkv/mkv_format.c [hunk 5]
```diff
@@ -808,10 +808,10 @@ mkv_metadata_parse(
 			return VOD_BAD_DATA;
 		}
 
-		if (metadata->base.tracks.nelts >= MAX_TRACK_COUNT)
+		if (metadata->base.tracks.nelts > MAX_TRACK_COUNT)
 		{
 			vod_log_error(VOD_LOG_ERR, request_context->log, 0,
-				"mkv_metadata_parse: track count exceeded the limit");
+				"mkv_metadata_parse: track count exceeded the limit of %i", MAX_TRACK_COUNT);
```
Comment (erankor) [inline vod/mkv/mkv_format.c:814]: may be better to add an explicit cast to ngx_int_t here

File: vod/mp4/mp4_parser.c [hunk 6]
```diff
@@ -2927,10 +2926,10 @@ mp4_parser_process_moov_atom_callback(void* ctx, atom_info_t* atom_info)
 	}
 
 	// add to the result array
-	if (result->base.tracks.nelts >= MAX_TRACK_COUNT)
+	if (result->base.tracks.nelts > MAX_TRACK_COUNT)
 	{
 		vod_log_error(VOD_LOG_ERR, context->request_context->log, 0,
-			"mp4_parser_process_moov_atom_callback: track count exceeded the limit");
+			"mp4_parser_process_moov_atom_callback: track count exceeded the limit of %i", MAX_TRACK_COUNT);
```
Comment (erankor) [inline vod/mp4/mp4_parser.c:2932]: cast also here

**Review by erankor — CHANGES_REQUESTED** [review]

File: vod/common.h [hunk 7]
```diff
@@ -16,8 +16,15 @@
...
-#define vod_set_bit(mask, index) (mask)[(index) >> 3] |= 1 << ((index) & 7)
+// bit sets
+#define vod_number_of_bits(i) (sizeof(i) * 8)
+#define vod_number_of_residual_bits(i) ((i) % vod_number_of_bits(uint64_t))
+#define vod_array_length_for_bits(i) ((i) / vod_number_of_bits(uint64_t) + !!((i) % vod_number_of_bits(uint64_t)))
+#define vod_is_bit_set(mask, index) (!!((mask)[(index) / vod_number_of_bits((mask)[0])] & ((uint64_t)1 << ((index) % vod_number_of_bits((mask)[0])))))
+#define vod_set_bit(mask, index) ((mask)[(index) / vod_number_of_bits((mask)[0])] |= ((uint64_t)1 << ((index) % vod_number_of_bits((mask)[0]))))
+#define vod_reset_bit(mask, index) ((mask)[(index) / vod_number_of_bits((mask)[0])] &= ~((uint64_t)1 << ((index) % vod_number_of_bits((mask)[0]))))
+#define vod_set_bits(mask, max_bits) vod_memset((mask), 0xff, sizeof(uint64_t) * vod_array_length_for_bits(max_bits));
+#define vod_reset_bits(mask, max_bits) vod_memzero((mask), sizeof(uint64_t) * vod_array_length_for_bits(max_bits));
```
[the remark this answers is not in the document]
Comment (JoelLinn) [reply vod/common.h:27]: True, these aren't used with the language mask or uint64_t*. Still, this might become an issue in the future, it would compile fine but be broken at runtime.

Are you willing to risk this?

File: vod/manifest_utils.c [hunk 8]
```diff
@@ -48,12 +48,34 @@ manifest_utils_write_bitmask(u_char* p, uint32_t bitmask, u_char letter)
...
+{
+	uint32_t i;
+	uint32_t count = vod_get_number_of_set_bits_in_mask(mask_temp, max_bits);
+
+	vod_memcpy(mask_temp, bitmask, sizeof(bitmask[0]) * vod_array_length_for_bits(max_bits));
+
+	while (count--)
+	{
+		i = vod_get_lowest_bit_set(mask_temp, max_bits);
```
[the remark this answers is not in the document]
Comment (JoelLinn) [reply vod/manifest_utils.c:61]: Yes I will change the return type of `vod_get_lowest_bit_set` to int32_t and have it return -1. Every value in the array needs to be compared to 0 anyways...

On a side-note:
https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html
> Built-in Function: int __builtin_ctz (unsigned int x)
>     Returns the number of trailing 0-bits in x, starting at the least significant bit position. **If x is 0, the result is undefined.**

File: vod/common.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (erankor) [reply vod/common.h:27]: another option is create macros specifically for track_mask_t (in media_format.h) and use explicitly sizeof(track_mask_t)

[the remark this answers is not in the document]
Comment (JoelLinn) [reply vod/common.h:27]: ok, so we would end up with 
```
vod_track_mask_set_bits
vod_track_mask_reset_bits
vod_track_mask_is_any_bit_set
```
etc...

[the remark this answers is not in the document]
Comment (erankor) [reply vod/common.h:27]: right, only the macros that care about the size.
it sounds ok to me, it's like we consider the track mask an object that has a set of methods associated with it

File: vod/manifest_utils.c (hunk 1 above)
Comment (JoelLinn) [reply vod/manifest_utils.c:143]: Nice idea but stringification will include any parentheses and other math expressions https://godbolt.org/z/v5bb9ae1r
`gcc -E` shows whats happening:
```
int sadfasfd() {
    return (sizeof("(12345)") - 1);
}

```

Comment (JoelLinn) [reply vod/manifest_utils.c:143]: Unpopular opinion (not proud): 
```C
#define MAX_TRACK_INDEX_LEN ((int)log10(MAX_TRACK_COUNT) + 1)
```
GCC will optimise this out even with no optimization enabled.
Clang is fine beginning with `-O1`

Comment (erankor) [reply vod/manifest_utils.c:143]: Yeah, I know, I thought about the parenthesis, I'm ok with that, don't like the log10 solution :)

Comment (JoelLinn) [reply vod/manifest_utils.c:143]: Ok, I will add a comment that the value is slightly off

**Comment by JoelLinn** [comment]
Integrated the requested changes, I hope I didn't miss anything.

File: vod/manifest_utils.c [hunk 9]
```diff
@@ -48,12 +48,41 @@ manifest_utils_write_bitmask(u_char* p, uint32_t bitmask, u_char letter)
...
 
-static uint32_t*
+static u_char*
+manifest_utils_write_bitmask(u_char* p, uint64_t* bitmask, uint64_t* mask_temp, uint32_t max_bits, u_char letter)
+{
+	uint32_t i;
+
+	vod_memcpy(mask_temp, bitmask, sizeof(bitmask[0]) * vod_array_length_for_bits(max_bits));
+
+	i = vod_get_lowest_bit_set(mask_temp, max_bits);
```
Comment (erankor) [inline vod/manifest_utils.c:58]: i'd prefer to avoid the two lines that call vod_get_lowest_bit_set, and use -
```
for (;;)
{
    i = vod_get_lowest_bit_set(mask_temp, max_bits);
    if (i < 0)
    {
        break;
    }
    ...
```

Comment (erankor) [reply vod/manifest_utils.c:58]: also, i think there's a bug here - with the while impl, it should be >= 0, not > 0

File: vod/manifest_utils.c [hunk 10]
```diff
@@ -244,29 +263,29 @@ manifest_utils_build_request_params_string(
...
 	// video tracks
-	if (tracks_mask[MEDIA_TYPE_VIDEO] == 0xffffffff)
+	if (vod_track_mask_are_all_bits_set(tracks_mask[MEDIA_TYPE_VIDEO]))
 	{
 		result_size += sizeof("-v0") - 1;
 	}
 	else
 	{
-		result_size += vod_get_number_of_set_bits(tracks_mask[MEDIA_TYPE_VIDEO]) * (sizeof("-v32") - 1);
+		result_size += vod_track_mask_get_number_of_set_bits(tracks_mask[MEDIA_TYPE_VIDEO]) * (sizeof("-v999") - 1);
```
Comment (erankor) [inline vod/manifest_utils.c:276]: need to use MAX_TRACK_INDEX_LEN here

File: vod/manifest_utils.c [hunk 11]
```diff
@@ -244,29 +263,29 @@ manifest_utils_build_request_params_string(
...
 	// audio tracks
-	if (tracks_mask[MEDIA_TYPE_AUDIO] == 0xffffffff)
+	if (vod_track_mask_are_all_bits_set(tracks_mask[MEDIA_TYPE_AUDIO]))
 	{
 		result_size += sizeof("-a0") - 1;
 	}
 	else
 	{
-		result_size += vod_get_number_of_set_bits(tracks_mask[MEDIA_TYPE_AUDIO]) * (sizeof("-a32") - 1);
+		result_size += vod_track_mask_get_number_of_set_bits(tracks_mask[MEDIA_TYPE_AUDIO]) * (sizeof("-a999") - 1);
```
Comment (erankor) [inline vod/manifest_utils.c:286]: here too

File: vod/manifest_utils.h [hunk 12]
```diff
@@ -46,12 +46,12 @@ typedef struct {
 // functions
 vod_status_t manifest_utils_build_request_params_string(
 	request_context_t* request_context,
-	uint32_t* has_tracks,
+	track_mask_t has_tracks[MEDIA_TYPE_COUNT],
```
Comment (erankor) [inline vod/manifest_utils.h:49]: i prefer track_mask_t* instead of []'s (also, that's how it's defined in the C file)

File: vod/manifest_utils.h [hunk 13]
```diff
@@ -46,12 +46,12 @@ typedef struct {
...
 	uint32_t segment_index,
 	uint32_t sequences_mask,
 	sequence_tracks_mask_t* sequence_tracks_mask,
 	sequence_tracks_mask_t* sequence_tracks_mask_end,
-	uint32_t* tracks_mask,
+	track_mask_t tracks_mask[MEDIA_TYPE_COUNT],
```
Comment (erankor) [inline vod/manifest_utils.h:54]: same here

File: vod/media_format.h [hunk 14]
```diff
@@ -14,7 +14,16 @@
...
+#ifdef NGX_VOD_MAX_TRACK_COUNT
+#if NGX_VOD_MAX_TRACK_COUNT % 64 != 0
+#error MAX_TRACK_COUNT must be a multiple of 64!
+#endif
+#define MAX_TRACK_COUNT (NGX_VOD_MAX_TRACK_COUNT)
+#else
+#define MAX_TRACK_COUNT (64)
+#endif
+// The parentheses in MAX_TRACK_COUNT are included in the stringification
+#define MAX_TRACK_INDEX_LEN (sizeof(ngx_value(MAX_TRACK_COUNT)) - 2 - 1)
```
Comment (erankor) [inline vod/media_format.h:26]: would prefer to put it inside the if/else, and use NGX_VOD_MAX_TRACK_COUNT/hardcoded value of 2, instead of doing -2

File: vod/media_format.h [hunk 15]
```diff
@@ -374,4 +386,13 @@ vod_status_t media_format_finalize_track(
...
 	media_info_t* media_info);
 
+
+#define vod_track_mask_set_all_bits(mask) vod_set_all_bits((mask), MAX_TRACK_COUNT)
+#define vod_track_mask_reset_all_bits(mask) vod_reset_all_bits((mask), MAX_TRACK_COUNT)
+#define vod_track_mask_get_number_of_set_bits(mask) vod_get_number_of_set_bits_in_mask((mask), MAX_TRACK_COUNT)
+#define vod_track_mask_are_all_bits_set(mask) vod_are_all_bits_set((mask), MAX_TRACK_COUNT)
+#define vod_track_mask_is_any_bit_set(mask) vod_is_any_bit_set((mask), MAX_TRACK_COUNT)
+#define vod_track_mask_get_lowest_bit_set(mask) vod_get_lowest_bit_set((mask), MAX_TRACK_COUNT)
+#define vod_track_mask_and_bits(dst, a, b) vod_and_bits((dst), (a), (b), MAX_TRACK_COUNT)
```
Comment (erankor) [inline vod/media_format.h:396]: please move to the 'macros' section in the beginning of the file

**Review by erankor — CHANGES_REQUESTED** [review]
@JoelLinn added a few more comments, after these are fixed, will need to run regression tests, and then we should be able to merge

**Comment by JoelLinn** [comment]
Pushed changes.

**Comment by JoelLinn** [comment]
> @JoelLinn added a few more comments, after these are fixed, will need to run regression tests, and then we should be able to merge

Any results from the regression tests yet?

File: vod/media_set_parser.c [hunk 16]
```diff
@@ -648,10 +648,10 @@ media_set_parse_source(
 	context->base.sources_head = source;
 
 	vod_log_debug4(VOD_LOG_DEBUG_LEVEL, context->base.request_context->log, 0,
-		"media_set_parse_source: parsed clip source - path=%V tracks[v]=0x%uxD tracks[a]=0x%uxD, clipFrom=%uL", 
+		"media_set_parse_source: parsed clip source - path=%V tracks[v]=0x%uxL tracks[a]=0x%uxL, clipFrom=%uxL",
```
Comment (erankor) [inline vod/media_set_parser.c:651]: clipFrom should remain %uL (not hex)

File: vod/common.h [hunk 17]
```diff
@@ -16,13 +16,19 @@
 #define vod_div_ceil(x, y) (((x) + (y) - 1) / (y))
 #define vod_array_entries(x) (sizeof(x) / sizeof(x[0]))
 
-#define vod_is_bit_set(mask, index) (((mask)[(index) >> 3] >> ((index) & 7)) & 1)
-#define vod_set_bit(mask, index) (mask)[(index) >> 3] |= 1 << ((index) & 7)
+// bit sets
+#define vod_number_of_bits(i) (sizeof(i) * 8)
```
Comment (erankor) [inline vod/common.h:20]: looking at this again, since we already decided we use uint64_t for the bitmask elements, and we're making this assumption in a few place, I think it would be cleaner to just replace all vod_number_of_bits(...) with '64'

**Review by erankor — CHANGES_REQUESTED** [review]
@JoelLinn the regression tests completed successfully!
However, I reviewed the whole thing again now, and have a couple of small comments.
We're getting there... :)

**Comment by JoelLinn** [comment]
> @JoelLinn the regression tests completed successfully! However, I reviewed the whole thing again now, and have a couple of small comments. We're getting there... :)

Relieved to hear there were no regressions... 
I made the two small changes you suggested and pushed them.

**Comment by erankor** [comment]
Merged, thanks!

**Comment by JoelLinn** [comment]
Thanks for putting the work into the reviews. 👍
PRs that touch many files and existing functionality are strenuous to review.
